### DISCUSSION TITLE
Shrink key file or generate key file based on key file

### DISCUSSION THREAD

**Discussion by Redsandro** [keepassxreboot/keepassxc#8403] [Q&A] [answered]
Someone used a video file as a key file. I'm tasked with safekeeping a backup of this key file. Is there any way I can generate a new (smaller!) key file that works with the same database?

In a more advanced version of this question, it would be especially helpful if I could use a thumbnail from the video and add some metadata to make it have the same hash which if I understand correctly is all that matters to keepassxc. It would be less confusing for me to (some day) hand a key back that they recognize because of the thumbnail. But that's just icing. Any key that's not huge would help me out.

**Comment by droidmonkey** [comment, marked as answer]
Just go to Database -> Database Security and generate a fresh key file and assign it to the database. We absolutely do not recommend using huge files or non purpose built files as the content could change even without your awareness and break access to your database.

**Reply by Redsandro** [reply]
I understand, but could you humor me?

Situation: I don't have access to the database. The owner is in the hospital, and the family has the database without the key, and a third unknown person has the password. We were instructed to contact the family in case of emergency. I will keep the original key file safe, but I'd like to convert it to an easily transferable key.

Technically speaking, I believe the database is protected with `sha256(sha256(password) + sha256(key file))`. If I understand correctly, I'm looking for a `sha256(small generated file)` that is equal to `sha256(very big key file)`. Is this correct?

_Edit:_ So what I'm not able to figure out is the following: Either having the original key makes this possible, or it is impossible regardless of having the original key because sha256 is too long/complex.

**Reply by droidmonkey** [reply]
sha256(small generated file) that is equal to sha256(very big key file)

This is impossible unless you crack SHA-256...

**Reply by Redsandro** [reply]
Okay that's clear. Thank you. I didn't know (on a technical level) if having the original key would make this somehow possible.

(I don't know if this is already the case, but I hope keepassxc discourages people creating a new database from using big files or video files as key file in the future. :slightly_smiling_face: )

**Reply by phoerious** [reply]
Actually, this is possible by using the old XML key file format. Get the Base64-encoded hash of your file like this:

```
shasum -a 256 keyfile-large.key | cut -d " " -f 1 | xxd -r -p | base64
```

And then create a keyfile with the following contents:
```
<?xml version="1.0" encoding="utf-8"?>
<KeyFile>
	<Meta>
		<Version>1.00</Version>
	</Meta>
	<Key>
		<Data>HASH</Data>
	</Key>
</KeyFile>
```

where you replace `HASH` with the output of the previous command.

**Reply by droidmonkey** [reply]
Oh great point!! Sorry about that, wasn't thinking that far outside the box.

**Reply by Redsandro** [reply]
> this is possible by using the old XML key file format.

Wait, Really? It's this simple? Never heard of this, either it predates my introduction to Keepass or I've been chronically unaware. Is there any documented intention of ~~deprecating this key~~ _removing support for this XML_ format any time soon?

Either way, thank you! :+1:

**Reply by phoerious** [reply]
The XML format will most likely stay. The other legacy formats may go at some point.
